internal/provider/data_source_http.go

```diff
@@ -0,0 +1,312 @@
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/hashicorp/terraform-plugin-framework-validators/int64validator"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+var _ tfsdk.DataSourceType = (*httpDataSourceType)(nil)
+
+type httpDataSourceType struct{}
+
+func (d *httpDataSourceType) GetSchema(context.Context) (tfsdk.Schema, diag.Diagnostics) {
+	return tfsdk.Schema{
+		Description: `
+The ` + "`http`" + ` data source makes an HTTP GET request to the given URL and exports
+information about the response.
+
+The given URL may be either an ` + "`http`" + ` or ` + "`https`" + ` URL. At present this resource
+can only retrieve data from URLs that respond with ` + "`text/*`" + ` or
+` + "`application/json`" + ` content types, and expects the result to be UTF-8 encoded
+regardless of the returned content type header.
+
+~> **Important** Although ` + "`https`" + ` URLs can be used, there is currently no
+mechanism to authenticate the remote server except for general verification of
+the server certificate's chain of trust. Data retrieved from servers not under
+your control should be treated as untrustworthy.`,
+
+		Attributes: map[string]tfsdk.Attribute{
+			"url": {
+				Description: "The URL for the request. Supported schemes are `http` and `https`.",
+				Type:        types.StringType,
+				Required:    true,
+			},
+
+			"request_headers": {
+				Description: "A map of request header field names and values.",
+				Type: types.MapType{
+					ElemType: types.StringType,
+				},
+				Optional: true,
+			},
+
+			"request_timeout": {
+				Description: "The request timeout in milliseconds.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Validators: []tfsdk.AttributeValidator{
+					int64validator.AtLeast(1),
+				},
+			},
+
+			"retry": {
+				Description: "Retry request configuration.",
+				Attributes: tfsdk.SingleNestedAttributes(map[string]tfsdk.Attribute{
```

Comment:
One thing that sometimes helps is to introduce a delay between attempts - maybe an exponential-back-off?. Similarly proposed here: https://github.com/hashicorp/terraform-provider-http/pull/87

Reply:
The `retryablehttp` client has an [exponential backoff by default and a default minimum wait between retries of 1 sec](https://pkg.go.dev/github.com/hashicorp/go-retryablehttp#DefaultBackoff).

We could expose the `RetryWaitMin` (and `RetryWaitMax`)? One of the reasons for using a `SingleNestedAttribute` for `retry` configuration was because of the possibility of needing to expose additional configuration.

Reply:
Even better if it's easy to expose and it's nicely "tucked away" in the nested object